ios/PolyPodApp/CoreDataModels/CoreDataStack.swift

```diff
@@ -0,0 +1,130 @@
+import Foundation
+import CoreData
+
+final class CoreDataStack {
+    static let shared = CoreDataStack()
+
+    // MARK: - Private API
+
+    private let container: NSPersistentContainer
+    
+    /*
+     This is the context to be used to perform all operation.
+     
+     Note: It is optional, as persistent stores are not loaded at initialization, but only when protected data is available.
+           When protected data becomes unavailable, context will be saved and nullified, see protectedDataWillBecomeUnavailable.
+     */
+    private var context: NSManagedObjectContext?
+
+    private init() {
+        let modelName = "PolyPodModel"
+        let model = NSManagedObjectModel.with(name: modelName, in: Bundle(for: CoreDataStack.self))
+        container = NSPersistentContainer.load(from: NSPersistentContainer.defaultDirectoryURL().appendingPathComponent("\(modelName).sqlite"),
+                                               name: modelName,
+                                               model: model)
+    }
+    
+    // MARK: - Internal API
+
+    /*
+     Allows asking if protected data is available.
+    
+     Note: It is unfortunate that this is needed. We cannot just rely on data protection delegate methods
+           from app delegate. For example, protectedDataWillBecomeUnavailable will only be called if device is locked
+           while polypod is in FOREGROUND. Otherwise no events of having protected data unavailable are sent.
+           This closure will be used to imperatively ask if protected data is avaialble for use.
+     */
+    var isProtectedDataAvailable: ((@escaping (Bool) -> Void) -> Void)?
+    
+    /// Use this function to perform any CoreData operation on PolyPod model.
+    func perform(_ operation: @escaping (NSManagedObjectContext) -> Void,
+                 file: String = #file,
+                 line: Int = #line,
+                 function: String = #function) {
+        guard let context = context else {
+            Log.error("Invalid attempt to execute core data operation in \(function) from \(file) at \(line)")
+            return
+        }
+        
+        isProtectedDataAvailable? { available in
+            if available {
+                context.perform { operation(context) }
+            } else {
+                Log.error("Invalid attempt to execute core data operation in \(function) from \(file) at \(line)")
```

Comment:
Sure, rewriting to what you've suggested is an option. My reasoning is that if context is nil, there is no point in asking if protected data is available, in a way I am short-circuiting the logic here.

I think it will be great to add an assertion failure if `isProtectedDataAvailable` is `nil`, since it will be a developer error to inject it properly.

Reply:
That's fine by me as well, what really bothers me is the duplication I guess :D But that's nit picky stuff. The *more* important points are:

1. Throwing errors instead of simply logging them. I think this could help eliminate the duplication because we could log errors higher up in the chain, where they get caught (if we don't already do that in this case), _and_ we can just use `PodApiError` for this parameterised error message. No strong opinion on whether to pass in `function` and all that - I'd probably just leave it out since we'd get a stack trace that should tell us this stuff anyway.
2. Also checking `isProtectedDataAvailable` for `nil`, and throwing an error if that's the case.

Reply:
Updated:

- Failing to set isProtectedDataAvailable is a developer error, if for some reason someone makes a change for it to not be set properly, Xcode will throw an error as I have added `!` on isProtectedDataAvailable.
- Now operation will get either the context or the error. Thus the client can propagate the error further if needed.
- Consolidated in one place the check for context and protected data availability to remove duplication. Also error is logged in one place here, I don't want to spread out this log message in all places where `perform` is called.

Reply:
Hm, I guess returning a `Result` object is good enough - in my understanding of the code, any calls we have that try to use `perform` do actually `try` to `get()` the result, so that'd fail, triggering the respective catch block, _possibly_ (that I don't know) showing the original error. Even if the latter is not the case, good enough for now I'd argue!

We will need a more thought through / consistent error handling throughout IMHO, but that's out of scope.

Reply:
Yes, your reasoning is right.
In Swift code we should normally use `Result` type. If PolyIn will be changed to complete with a `Result` then we don't need to use anymore `do-catch`, we can chain/map results.

Alternatively we can consider using new `[async await](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html)` API.